parley/src/cursor_test.rs

```diff
@@ -0,0 +1,270 @@
+    /// [`Affinity::Upstream`].
+    ///
+    /// The needle must be unique in the text to avoid ambiguity.
+    ///
+    /// ### Panics
+    ///
+    /// - If the needle is not found in the text.
+    /// - If the needle is found multiple times in the text.
+    #[track_caller]
+    pub fn cursor_after(&self, needle: &str) -> Cursor {
+        let index = self.get_unique_index("cursor_after", needle);
+        let index = index + needle.len();
+        Cursor::from_byte_index(&self.layout, index, Affinity::Upstream)
+    }
+
+    fn cursor_to_monospace(&self, cursor: Cursor, is_correct: bool) -> String {
+        fn check_no_color() -> bool {
+            let Some(env_var) = std::env::var_os("NO_COLOR") else {
+                return false;
+            };
+            let env_var = env_var.to_str().unwrap_or_default().trim();
+
+            if env_var == "0" {
+                return false;
+            }
+            if env_var.to_ascii_lowercase() == "false" {
+                return false;
+            }
+            true
+        }
+
+        // NOTE: The background color doesn't carry important information,
+        // so we do a simple implementation, without worrying about
+        // color-blindness and platform issues.
+        let ansi_bg_color = if cfg!(not(unix)) || check_no_color() {
+            ""
+        } else if is_correct {
+            // Green background
+            "\x1b[48;5;70m"
+        } else {
+            // Red background
+            "\x1b[48;5;160m"
+        };
+        let ansi_reset = if cfg!(not(unix)) { "" } else { "\x1b[0m" };
+        let index = cursor.index();
+        let affinity = cursor.affinity();
+
+        let cursor_str = if affinity == Affinity::Upstream {
+            // - ANSI code for 'Set background color'
+            // - Unicode sequence for '▕' character
+            // - ANSI code for 'Reset all attributes'
+            format!("{ansi_bg_color}\u{2595}{ansi_reset}")
+        } else {
+            // - 1 space
+            // - ANSI code for 'Set background color'
+            // - Unicode sequence for '▏' character
+            // - ANSI code for 'Reset all attributes'
+            format!(" {ansi_bg_color}\u{258F}{ansi_reset}")
+        };
+
+        // FIXME - This assumes that the byte index of a string matches how many
+        // terminal tiles that string occupies. This is wrong for even trivial
+        // cases (eg unicode characters spanning multiple code points).
+        " ".repeat(index) + &cursor_str
+    }
+
+    #[track_caller]
+    fn cursor_assertion(&self, expected: Cursor, actual: Cursor) {
+        if expected == actual {
+            return;
+        }
+
+        // TODO - Check that the tested string doesn't include difficult
+        // characters (newlines, tabs, RTL text, etc.)
+        // If it does, we should still print the text on a best effort basis, but
+        // without visual cursors and with a warning that the text may not be accurate.
```

Comment:
As discussed on Zulip, the alternative/fallback is to do screenshots. I'd suggest that we should have screenshots unconditionally if this fails, and show the in-terminal version if there is no RTL region (we should strive to handle newlines properly, and we can't possibly detect non-tab extra-wide terminal characters so it's not worth trying)

This doesn't need to happen in this PR, but it might be worth leaving these breakcrumbs in the code